### PR TITLE
[SPARK-21723][ML] Fix writing LibSVM (key not found: numFeatures)

### PR DESCRIPTION
**Pull request by ProtD**
## What changes were proposed in this pull request?

Check the option "numFeatures" only when reading LibSVM, not when writing. When writing, Spark was raising an exception. After the change it will ignore the option completely. @liancheng @HyukjinKwon

(Maybe the usage should be forbidden when writing, in a major version change?).

## How was this patch tested?

Manual test, that loading and writing LibSVM files work fine, both with and without the numFeatures option.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by srowen** [comment]
Better title please?
See http://spark.apache.org/contributing.html
How does this work at all now?

**Review by WeichenXu123 — CHANGES_REQUESTED** [review]
Could you pls add a testcase for this fix ?

**Comment by ProtD** [comment]
@srowen It worked in v2.0, but was broken probably in v2.2.0 by b3d39620c563e5f6a32a4082aa3908e1009c17d2. Current unit tests check writing only for dataframes which were previously read from a LibSVM format, not general ones. (And I guess people don't write LibSVMs very often - that may be why nobody has reported it.)

@WeichenXu123 Yes, good idea, will do it!

**Comment by ProtD** [comment]
To reproduce the bug on v2.2 and v2.3:
```scala
import org.apache.spark.ml.linalg.Vectors
val rawData = Seq((1.0, Vectors.sparse(3, Seq((0, 2.0), (1, 3.0)))),
                  (4.0, Vectors.sparse(3, Seq((0, 5.0), (2, 6.0)))))
val dfTemp = spark.sparkContext.parallelize(rawData).toDF("label", "features")
dfTemp.coalesce(1).write.format("libsvm").save("...filename...")
```
This causes `java.util.NoSuchElementException: key not found: numFeatures`.

**Comment by ProtD** [comment]
I added the unit test, please review.

**Review by srowen — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
ok to test

File: mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala [hunk 1]
```diff
@@ -126,6 +130,29 @@ class LibSVMRelationSuite extends SparkFunSuite with MLlibTestSparkContext {
     }
   }
 
+  test("write libsvm data from scratch and read it again") {
+    val rawData = new java.util.ArrayList[Row]()
+    rawData.add(Row(1.0, Vectors.sparse(3, Seq((0, 2.0), (1, 3.0)))))
+    rawData.add(Row(4.0, Vectors.sparse(3, Seq((0, 5.0), (2, 6.0)))))
+
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala]: Subtle: it didn't like the whitespace on this line

Comment (ProtD) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala]: Fixed.

**Comment by srowen** [comment]
@ProtD this needs a JIRA or else needs to be linked to whatever one you opened, in the title

**Comment by ProtD** [comment]
@srowen Ok, I created and linked a JIRA.

File: mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala [hunk 2]
```diff
@@ -109,14 +112,15 @@ class LibSVMRelationSuite extends SparkFunSuite with MLlibTestSparkContext {
   test("write libsvm data and read it again") {
     val df = spark.read.format("libsvm").load(path)
     val tempDir2 = new File(tempDir, "read_write_test")
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala]: I suggest the temp dir name to be `Identifiable.randomUID("read_write_test")`. Avoid conflicts with other parallel running tests.

File: mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala [hunk 3]
```diff
@@ -109,14 +112,15 @@ class LibSVMRelationSuite extends SparkFunSuite with MLlibTestSparkContext {
...
-    val writepath = tempDir2.toURI.toString
+    val writePath = tempDir2.toURI.toString
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala]: use `tempDir2.getPath`

File: mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala [hunk 4]
```diff
@@ -109,14 +112,15 @@ class LibSVMRelationSuite extends SparkFunSuite with MLlibTestSparkContext {
...
     // TODO: Remove requirement to coalesce by supporting multiple reads.
-    df.coalesce(1).write.format("libsvm").mode(SaveMode.Overwrite).save(writepath)
+    df.coalesce(1).write.format("libsvm").mode(SaveMode.Overwrite).save(writePath)
 
-    val df2 = spark.read.format("libsvm").load(writepath)
+    val df2 = spark.read.format("libsvm").load(writePath)
     val row1 = df2.first()
     val v = row1.getAs[SparseVector](1)
     assert(v == Vectors.sparse(6, Seq((0, 1.0), (2, 2.0), (4, 3.0))))
+    Utils.deleteRecursively(tempDir2)
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala]: You can remove this cleanup I think. The test framework will clean temp dir automatically I think.

File: mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala (hunk 2 above)
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala]: Use `Utils.createTempDir`

Comment (ProtD) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala]: `Utils.createTempDir` seems to be a nicer way. The directory is automatically deleted when VM shuts down, so I believe no manual cleanup (cf. comment lower) is needed.

**Review by srowen — APPROVED** [review]

**Review by WeichenXu123 — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master/2.2

**Pull request closed without merging**